### ISSUE TITLE
Efficient flattening (and code generation) for large-scale network models

### ISSUE THREAD

**Issue by OpenModelica-TracImporter** [OpenModelica/OpenModelica#3678]

**Comment by perost** [comment]
Original ticket: https://trac.openmodelica.org/OpenModelica/ticket/3678

**Comment by perost** [comment]
@casella 

I've got something that's kind of working now in my [mergescalars branch](https://github.com/perost/OpenModelica/tree/mergescalars), but there's still some work left. The current implementation merges components in the model being instantiated if they:
* have the same scalar type (with the exact same name)
* are not themselves arrays (could probably be handled if we want)
* are not replaceable
* are not inner/outer
* are not conditional
* do not have redeclare modifiers or are themselves redeclare elements

Note that only components immediately in the model being instantiated will be merged, we don't merge subcomponents or inherited components since that would complicate the renaming a lot (doing it correctly would require instantiation, which kind of defeats the whole purpose). 

So for a model such as:
```Modelica
model A
  Real x;
  Real y;
end A;

model M
  A a1(x = 1, y(start = 0));
  A a2(y = 2);
  A a3(x = 3, y(start = 1) = 4);
equation
  a2.x = a1.x + a3.x;
end M;
```
we get:
```Modelica
class M
  Real $merged1[1].x = 1.0;
  Real $merged1[1].y(start = 0.0);
  Real $merged1[2].x;
  Real $merged1[2].y = 2.0;
  Real $merged1[3].x = 3.0;
  Real $merged1[3].y(start = 1.0) = 4.0;
equation
  $merged1[2].x = $merged1[1].x + $merged1[3].x;
end M;
```
There are some things left to fix though:
- [ ] It currently only works if the last component in a merge set has all modifiers, otherwise we get arrays with unknown type due to a quirk in how we type arrays in bindings.
- [ ] Something in the flattening phase scales very badly for bindings with large arrays, causing large models to grind to a halt during that phase.
- [ ] Every modifier is currently assumed to be non-final and non-each, which may not be true.
- [ ] It doesn't work with `-d=nonfScalarize` for some reason, it just fails somewhere.
- [ ] The new Absyn.Exp.EMPTY needs to be handled in the Absyn.Exp utility functions (printing, comparing, etc).
- [ ] Testcases

**Comment by casella** [comment]
>  The current implementation merges components in the model being instantiated if they:
> 
>     * have the same scalar type (with the exact same name)
> 
>     * are not themselves arrays (could probably be handled if we want)
> 
>     * are not replaceable
> 
>     * are not inner/outer
> 
>     * are not conditional
> 
>     * do not have redeclare modifiers or are themselves redeclare elements

Very good. This looks perfect for the application of large power grids. @adriguir, do you agree?

> Note that only components immediately in the model being instantiated will be merged, we don't merge subcomponents or inherited components since that would complicate the renaming a lot (doing it correctly would require instantiation, which kind of defeats the whole purpose).

Just to make sure I understand this correctly, consider this exemplary case
```modelica
model ControlledGenerator
  Generator g;
  Controller c;
  Filter f1;
  Filter f2;
  ...
end ControlledGenerator;

model Grid
  ControlledGenerator gc1(g(W = 1e9), c(droop = 0.05));
  ControlledGenerator gc2(g(W = 6e8), c(droop = 0.05));
  ControlledGenerator gc3(g(W = 2e8), c(droop = 0.02));
  ...
end Grid;
```
I understand what you mean is that `gc1`, `gc2`, `gc3` etc will be collected in an array `$merged[N]` (or `$ControlledGenerator[N]` according to my proposal), but you won't be doing the same for `f1` and `f2`. In other words, only top-level instances will be merged into arrays. Is that correct?

> ```modelica
> class M
>   Real $merged1[1].x = 1.0;
>   Real $merged1[1].y(start = 0.0);
>   Real $merged1[2].x;
>   Real $merged1[2].y = 2.0;
>   Real $merged1[3].x = 3.0;
>   Real $merged1[3].y(start = 1.0) = 4.0;
> equation
>   $merged1[2].x = $merged1[1].x + $merged1[3].x;
> end M;
> ```

The first problem that I see here is that this is not legal Modelica, because you have separate declarations of array elements. I would need to get declarations such as

```modelica
class M
  Real $merged1.x[3];
  Real $merged1.y[3];
  Real $merged1.z[3];
   ...
end M;
```
Of course the question then is: how do we manage attribute modifiers and binding equations, if they are not applied to the same parameters and variables in each instance?

For parameters, we could apply the default value for those elements that do not have an explicit modifier, so you can build full-sized array modifiers and then apply them wholesale to the array declaration. That would be very useful. 

For variables, I would say that either the modification pattern is the same for all instances (and then you generate a bulk array modifier or binding equation for the array declaration), otherwise you give up. I guess this should be good enough to handle the cases of our interest.

Last, but not least, I would prefer to use the class name for the merged variables, e.g. `$A1` in your example, so that there is some hint to where these variables come from - instance names are lost by definition, but at least their type isn't. 

> There are some things left to fix though:
> 
>     * [ ]  It currently only works if the last component in a merge set has all modifiers, otherwise we get arrays with unknown type due to a quirk in how we type arrays in bindings.

See above

>     * [ ]  Something in the flattening phase scales very badly for bindings with large arrays, causing large models to grind to a halt during that phase.

Ouch. How large is large?  A few hundred, a few thousand, or a few million?

>     * [ ]  Every modifier is currently assumed to be non-final and non-each, which may not be true.

See above, either they all have final/each applied to the same modifiers, or you give up.

>     * [ ]  It doesn't work with `-d=nonfScalarize` for some reason, it just fails somewhere.

I'm afraid that defeats the whole purpose of this proposal. If you flatten to scalar declarations and equations, having `a1` and `a2` vs `$merged[1]` and `$merged[2]` makes little difference. With -f, you'll get `'a1'` and `'a2'` vs. `'$merged[1]'` and `'$merged[2]'`, but it doesn't make any difference, they're just to strings identifying two scalar Reals. 
 
>     * [ ]  The new Absyn.Exp.EMPTY needs to be handled in the Absyn.Exp utility functions (printing, comparing, etc).
Maybe this could wait for the time being.
>     * [ ]  Testcases
We have several of them in our benchmark repo, which is still private on GitHub, did I give you access to them?

**Comment by perost** [comment]
> Just to make sure I understand this correctly, consider this exemplary case
> 
> ```modelica
> model ControlledGenerator
>   Generator g;
>   Controller c;
>   Filter f1;
>   Filter f2;
>   ...
> end ControlledGenerator;
> 
> model Grid
>   ControlledGenerator gc1(g(W = 1e9), c(droop = 0.05));
>   ControlledGenerator gc2(g(W = 6e8), c(droop = 0.05));
>   ControlledGenerator gc3(g(W = 2e8), c(droop = 0.02));
>   ...
> end Grid;
> ```
> 
> I understand what you mean is that `gc1`, `gc2`, `gc3` etc will be collected in an array `$merged[N]` (or `$ControlledGenerator[N]` according to my proposal), but you won't be doing the same for `f1` and `f2`. In other words, only top-level instances will be merged into arrays. Is that correct?

Correct. Otherwise we could have references to `f1` and `f2` in `Grid`, and then the renaming becomes far more complicated. If we restrict it to top-level instances it's trivial, since we know that e.g. the name `gc1` can only refer to the component `gc1`.

> The first problem that I see here is that this is not legal Modelica, because you have separate declarations of array elements. I would need to get declarations such as

That's the flat output from the compiler, not how it looks like after the transformation. Currently I do the transformation between looking up the class and instantiating it in the NF, but it's done on the SCode so we could also have an API call for it in case you also want to save the model.

> Of course the question then is: how do we manage attribute modifiers and binding equations, if they are not applied to the same parameters and variables in each instance?
> 
> For parameters, we could apply the default value for those elements that do not have an explicit modifier, so you can build full-sized array modifiers and then apply them wholesale to the array declaration. That would be very useful.
> 
> For variables, I would say that either the modification pattern is the same for all instances (and then you generate a bulk array modifier or binding equation for the array declaration), otherwise you give up. I guess this should be good enough to handle the cases of our interest.

I implemented it so that modifiers that are missing are simply ignored, but now I realize that they're actually going to override any existing modifiers first and then be removed. So I guess that's another thing to add to the todo list.

> Last, but not least, I would prefer to use the class name for the merged variables, e.g. `$A1` in your example, so that there is some hint to where these variables come from - instance names are lost by definition, but at least their type isn't.

Sounds like a good idea.

> > * [ ]  Something in the flattening phase scales very badly for bindings with large arrays, causing large models to grind to a halt during that phase.
> Ouch. How large is large? A few hundred, a few thousand, or a few million?

The `SystemLarge` model with 5000 instances takes about 8 sec for the flattening phase (compared to 0.6s without merging), while 50000 instances takes longer than I care to wait. I think the issue is that our array expressions are not arrays but linked lists, so accessing an array element is linear time instead of constant. This is the cause of several performance issues we have, and linked lists are just generally bad for performance due to poor cache locality. But changing array expressions to be actual arrays is a pretty huge task, and arrays are also not as well supported in MetaModelica.

Since we probably don't want to scalarize these models anyway I will try to fix `-d=nonfScalarize` first, and maybe that will be good enough for now.

> > ```
> > * [ ]  The new Absyn.Exp.EMPTY needs to be handled in the Absyn.Exp utility functions (printing, comparing, etc).
> > ```
> 
> Maybe this could wait for the time being.

Not really, adding new types without proper support isn't very nice. It's also one of the easier things to do, I just added it to the list so I don't forget.

> 
> > ```
> > * [ ]  Testcases
> > ```
> 
> We have several of them in our benchmark repo, which is still private on GitHub, did I give you access to them?

I don't know, but I will add my own test cases anyway to make sure I cover the corner cases.

**Comment by casella** [comment]
> Correct. Otherwise we could have references to `f1` and `f2` in `Grid`, and then the renaming becomes far more complicated. If we restrict it to top-level instances it's trivial, since we know that e.g. the name `gc1` can only refer to the component `gc1`.

Deal. Maybe in the future we could extend it slightly, like, instead of only merging top level instances, do merge instances within certain instantiated models that are listed somewhere (e.g. in an annotation9.

> The `SystemLarge` model with 5000 instances takes about 8 sec for the flattening phase (compared to 0.6s without merging), while 50000 instances takes longer than I care to wait. I think the issue is that our array expressions are not arrays but linked lists, so accessing an array element is linear time instead of constant. This is the cause of several performance issues we have, and linked lists are just generally bad for performance due to poor cache locality. But changing array expressions to be actual arrays is a pretty huge task, and arrays are also not as well supported in MetaModelica.
> 
> Since we probably don't want to scalarize these models anyway I will try to fix `-d=nonfScalarize` first, and maybe that will be good enough for now.

Sounds good. I guess 5000 instances is good enough already, even though very large EU-level models could have some components reaching the 10000 or even 20000 mark. That would be a huge improvement over the current state of the art for such models, which take 30+ minutes to compile with scalarized flattening and codegen. If this looks promising, then we can get proper funding to do this with arrays. Or, alternatively, write pre-processors that do this in other languages, e.g. C++, Python or Julia, which have better support for array data structures.

> Not really, adding new types without proper support isn't very nice. It's also one of the easier things to do, I just added it to the list so I don't forget.

OK.

> > We have several of them in our benchmark repo, which is still private on GitHub, did I give you access to them?
> 
> I don't know, but I will add my own test cases anyway to make sure I cover the corner cases.

I'll get back to you by email with that.

**Comment by casella** [comment]
Here it is, no need to keep it confidential

[TestArray.mo.txt](https://github.com/OpenModelica/OpenModelica/files/7961854/TestArray.mo.txt)

See sub-package Vectorizable. We currently flatten with flags 
```
-f -d=nonfScalarize,arrayConnect,combineSubscripts,printRecordTypes --newBackend --showStructuralAnnotations`
```

**Comment by perost** [comment]
I've now changed the grouping of components such that only components with the same modifiers are merged together. So for e.g.:
```Modelica
ResistorSource rs_1(R = 1, u = sin(time));
ResistorSource rs_2(R = 2);
ResistorSource rs_3(R = 3, u = sin(time));
ResistorSource rs_4(R = 4);
ResistorSource rs_5(R = 5, u = sin(time)));
```
we get:
```Modelica
ResistorSource $ResistorSource1(R = {1, 3, 5}, u = {sin(time), sin(time), sin(time)});
ResistorSource $ResistorSource2(R = {1, 4});
```
This gets rid of the need to have "empty" expressions in the arrays, which solves the issues we had with nonfScalarize, bad typing of binding arrays, and overwriting modifiers with missing modifiers. We get potentially less merging if there are many instances with different modifier sets, but that's probably a cost worth paying for not having to deal with all the issues we otherwise need to solve.

We could also easily turn e.g. `u = {sin(time), sin(time), sin(time)}` into `each u = sin(time)`, but I'm not sure if that's a common enough use case to justify the extra cost of checking if all arrays elements are equal in modifier bindings.

**Comment by perost** [comment]
Here are some performance numbers using the `SystemLarge` model from the original ticket with varying amounts of resistor instances. `-d=nonfScalarize` was used in all the cases.

| Instances | No merging | Merging | Merging + arrayConnect | 
| --------: | ---------: | ------: | ---------------------: |
| 50        | 0.04s       | 0.007s    | 35s                    |
| 5,000     | 4.4s       | 1.1s    | X                      |
| 50,000    | 43s        | 5.2s    | X                      |
| 500,000   | X          | 56s     | X                      |

The first thing to notice is that the array connection handling scales awfully bad and has big issues already with 50 instances. I tried with 500 instances but gave up on that. I don't know why it scales so badly, but the issue seems to be the set-based graph connected components algorithm. If we want to use this I think we'd have to merge connections too somehow.

Using the normal connection handling there's no problem handling half a million instances though, but then we get scalarized equations of course. The scaling when using merging gets worse the larger the model is due to this, but even with half a million connections the connection handling only takes about 15s.

The reason why there's an X in the last row of the no merging column is because I ran out of memory (I only had ~34GB free ...) and the process was killed. But if I did have enough memory it would take about 7 min, given that it continues to scale linearly.

**Comment by casella** [comment]
@perost, I don't get why merging takes less time than no mergin, could you explain?

**Comment by casella** [comment]
> The first thing to notice is that the array connection handling scales awfully bad and has big issues already with 50 instances. I tried with 500 instances but gave up on that. I don't know why it scales so badly, but the issue seems to be the set-based graph connected components algorithm. If we want to use this I think we'd have to merge connections too somehow.

In fact, I think we really don't care. The example in the original ticket (and also the examples in the ScalablePowerGrids benchmark) is somewhat artificial, because it has a very regular connection structure. In real power systems this is not the case, there are no rectangular connection patterns, so there is no chance (hence no need) of collecting connect statements in for loops. So, you can just skip that part.

 What is really important is to exploit the repetitiveness of the 1000s of instances of the same models, each of which may have 100s of equations. We can manage the 1000s of connect statements one by one; in fact, we have to, because they follow no regular pattern that can be recast as an array.

> Using the normal connection handling there's no problem handling half a million instances though

Good, that's exactly what we need 😃

**Comment by perost** [comment]
> @perost, I don't get why merging takes less time than no mergin, could you explain?

The time is the total time to flatten the model, so merging scalars into arrays is faster in this case since it makes the later phases faster when using `-d=nonfScalarize`. Which is exactly what we're aiming for after all.

The merging itself takes about 6s for the 500,000 case and seems to scale pretty much linearly.

**Comment by casella** [comment]
OK. It seems that if you forgo useless connection merging we get the perfect combination.

**Comment by perost** [comment]
I've now opened #8490 that adds the `-d=mergeComponents` flag to enable this feature, please try it out and check how it works for you.

And just to summarize: the feature should be used with `-d=nonfScalarize` but not `-d=arrayConnect`, otherwise you'll get scaling issues with large models. Other flags shouldn't matter.

**Comment by casella** [comment]
> I've now opened #8490 that adds the `-d=mergeComponents` flag to enable this feature, please try it out and check how it works for you.

OK, I'll try it tomorrow morning

> And just to summarize: the feature should be used with `-d=nonfScalarize` but not `-d=arrayConnect`, otherwise you'll get scaling issues with large models. Other flags shouldn't matter.

I'm a bit confused by this. The idea of -d=arrayConnect was to keep equations coming from array (or for-loop) connect statements together as array (or for-loop) equations. This is different from collecting scalar connection equations into arrays. Can't we have both?

This is actually no big deal, because depending on the type of model, you either need one feature or the other. But it is a bit inconvenient that you have to set up a debug flag to get the correct behaviour.

**Comment by perost** [comment]
> > And just to summarize: the feature should be used with `-d=nonfScalarize` but not `-d=arrayConnect`, otherwise you'll get scaling issues with large models. Other flags shouldn't matter.
> 
> I'm a bit confused by this. The idea of -d=arrayConnect was to keep equations coming from array (or for-loop) connect statements together as array (or for-loop) equations. This is different from collecting scalar connection equations into arrays. Can't we have both?
> 
> This is actually no big deal, because depending on the type of model, you either need one feature or the other. But it is a bit inconvenient that you have to set up a debug flag to get the correct behaviour.

I'm not sure what you mean, but the issue is that `-d=arrayConnect` can't handle large amounts of connect equations (or possibly it's the structure in the `SystemLarge` model that causes issues for the algorithm). And the only feature we have that potentially collects scalar connect equations into arrays is `-d=arrayConnect`.

**Comment by casella** [comment]
@perost, I guess I got lost in the story of non-scalarized connect support in the NF. Let me try to summarize the story, to see if we're on the same page.

In the beginning there was darkness, i.e., connect statements in for loops were simply ignored when the system was flattened with `-d=nonfScalarize`, see [#5643](https://trac.openmodelica.org/OpenModelica/ticket/5643). Then came some light with [PR 6848](https://github.com/OpenModelica/OpenModelica/pull/6848) and [PR 6863](https://github.com/OpenModelica/OpenModelica/pull/6848). I'm not sure how those PRs work, but based on what I read here, I assume that you first unroll the connect statements, handle them on a one-by-one basis, and then merge them together to get array equations back. This allows O(1) complexity in the backend and generated code, at the expense of at least O(N) complexity in the frontend, which is not ideal but still be OK if the elementary operations are fast.

Is that correct?

I also made a proposal in [#5677](https://trac.openmodelica.org/OpenModelica/ticket/5677) to handle some special cases of trivial connect statements, that result in `c1.e = c2.e` and `c1.f + c2.f = 0` equation pairs, with O(1) complexity. That was eventually abandoned since #5643 covered that case, albeit less efficiently. The situation is fine for me at the moment, though I guess we could still revive that proposal if we find that we have real life cases where the performance is bad because of large numbers of trivial connect statements, so we really need need to handle them with O(1) complexity.

Current NF performance issues with the O(N) `-d=arrayConnect` can of course be circumvented by not activating that flag. What I am wondering is whether it is possible for the NF to understand the difference between connect equations coming from array or for loops in the original model, and connect equations coming from the reconstructed arrays that are the subject of this ticket. The former are always worth mergining, even if it takes time to do that, the latter usually are not, so I would recommend that -d=arrayConnect skips merging them.

**Comment by perost** [comment]
> In the beginning there was darkness, i.e., connect statements in for loops were simply ignored when the system was flattened with `-d=nonfScalarize`, see [#5643](https://trac.openmodelica.org/OpenModelica/ticket/5643). Then came some light with [PR 6848](https://github.com/OpenModelica/OpenModelica/pull/6848) and [PR 6863](https://github.com/OpenModelica/OpenModelica/pull/6848). I'm not sure how those PRs work, but based on what I read here, I assume that you first unroll the connect statements, handle them on a one-by-one basis, and then merge them together to get array equations back. This allows O(1) complexity in the backend and generated code, at the expense of at least O(N) complexity in the frontend, which is not ideal but still be OK if the elementary operations are fast.
> 
> Is that correct?

#6848 and #6863 have nothing to do with `-d=nonfScalarize`, they are just some bug fixes for `-d=arrayConnect`. The issue with connect equations when using `d=nonfScalarize` that I mention in [my comment on #5643](https://trac.openmodelica.org/OpenModelica/ticket/5643#comment:5) is still a problem, i.e. array connectors are not scalarized properly which causes issues for the normal connection handling which assumes everything is scalarized (this is not an issue for the `SystemLarge` model though since it only has scalar connectors).

> I also made a proposal in [#5677](https://trac.openmodelica.org/OpenModelica/ticket/5677) to handle some special cases of trivial connect statements, that result in `c1.e = c2.e` and `c1.f + c2.f = 0` equation pairs, with O(1) complexity. That was eventually abandoned since #5643 covered that case, albeit less efficiently. The situation is fine for me at the moment, though I guess we could still revive that proposal if we find that we have real life cases where the performance is bad because of large numbers of trivial connect statements, so we really need need to handle them with O(1) complexity.
> 
> Current NF performance issues with the O(N) `-d=arrayConnect` can of course be circumvented by not activating that flag. What I am wondering is whether it is possible for the NF to understand the difference between connect equations coming from array or for loops in the original model, and connect equations coming from the reconstructed arrays that are the subject of this ticket. The former are always worth mergining, even if it takes time to do that, the latter usually are not, so I would recommend that -d=arrayConnect skips merging them.

The merging of connect equations when using `-d=arrayConnect` is a side-effect of the set-based graphs algorithm used, I wouldn't even know where to begin to selectively merge only some connect equations.

**Comment by casella** [comment]
OK, sorry, now I'm even more confused about what -d=arrayConnect does. The [documentation](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-debug-arrayconnect) is a classic tautology and doesn't help much:
> Use experimental array connection handler.
@perost, could you explain what this module does in one or two lines, and add them to the documentation.

**Comment by casella** [comment]
> The merging of connect equations when using `-d=arrayConnect` is a side-effect of the set-based graphs algorithm used, I wouldn't even know where to begin to selectively merge only some connect equations.

I understand you have a bunch of scalar connect equations, involving connector names. Some of them can be recognized as coming from the array-fication of bunches of independent instances of the same class, because of their prefix or something. Can't you just preliminarly remove them from the set to which the graph algorithm is applied?

**Comment by perost** [comment]
> OK, sorry, now I'm even more confused about what -d=arrayConnect does. The [documentation](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html#omcflag-debug-arrayconnect) is a classic tautology and doesn't help much:
> 
> > Use experimental array connection handler.
> 
> @perost, could you explain what this module does in one or two lines, and add them to the documentation.

`-d=arrayConnect` is an alternative connection handler based on the set-based graphs algorithm from Ernesto Kofman's group. I could add that it's based on set-based graphs to the flag description, but I'm not sure that clarifies much.

> I understand you have a bunch of scalar connect equations, involving connector names. Some of them can be recognized as coming from the array-fication of bunches of independent instances of the same class, because of their prefix or something. Can't you just preliminarly remove them from the set to which the graph algorithm is applied?

I'm not sure I follow, you want to use the normal connection handling on connect equations involving merged components and the array connection handling on other connect equations?

**Comment by casella** [comment]
> `-d=arrayConnect` is an alternative connection handler based on the set-based graphs algorithm from Ernesto Kofman's group. I could add that it's based on set-based graphs to the flag description, but I'm not sure that clarifies much.

What is not clear to me is what it does (how it does it, i.e. usig set-based graphs is also interesting, but that's not my point). I reckon it does something with arrays of connect statements, but what is it? Try to process the without unrolling them as much as possible? Unrolling them first, and then trying to merge them back? Something like this.

> I'm not sure I follow, you want to use the normal connection handling on connect equations involving merged components and the array connection handling on other connect equations?

Yes, that's the idea. At least optionally (yes, that's another debug flag...). My argument is that merged components normally have an irregular connection structure, otherwise you would create them with for loops and arrays in the first place. Hence, there's no point trying to merge them back into arrays, because they may be fragmented, and in general not very useful.

Conversely, if the original model has for loops, chances are that merging them can be highly successful and lead to much faster backend analysis and smaller runtime code.

**Comment by perost** [comment]
> > `-d=arrayConnect` is an alternative connection handler based on the set-based graphs algorithm from Ernesto Kofman's group. I could add that it's based on set-based graphs to the flag description, but I'm not sure that clarifies much.
> 
> What is not clear to me is what it does (how it does it, i.e. using set-based graphs is also interesting, but that's not my point). I reckon it does something with arrays of connect statements, but what is it? Try to process the without unrolling them as much as possible? Unrolling them first, and then trying to merge them back? Something like this.
> 
> Also, clarifying the relationship with -nonfScalarize would be useful, that's also not too clear to me. I understand this flag is only meaningful if you avoid scalarization, but in some comment you wrote they are independent, so I was a bit lost.

The algorithm is based on ranges, so if you have two connector arrays c1[100] and c2[50] they'll be represented as e.g. 1:100 and 101:150. Then the algorithm splits and merges algorithms based on the connections somehow, and you end up with the sets of ranges that are connected. I don't know the details, but there's a [paper on it](https://arxiv.org/abs/2008.04183) if you're interested.

And the relationship between the flags is just that if you don't use `nonfScalarize` then everything will be scalarized, so there won't be any arrays for the algorithm to exploit. So you most likely want to use the flags together, but there's no actual dependency between them.

**Comment by casella** [comment]
@perost, I tried the models in `TestArray`, attached in a previous comment, with flags `-f -d=nonfScalarize,mergeComponents`. Model `TestArray.Vectorizable.B` was flattened as expected, but `TestArray.Vectorizable.C` gave:
> Translation Error
> Internal error Instantiation of TestArray.Vectorizable.C failed with no error message.

I guess there is still something going wrong with the mergin. Can you please check?

**Comment by perost** [comment]
> @perost, I tried the models in `TestArray`, attached in a previous comment, with flags `-f -d=nonfScalarize,mergeComponents`. Model `TestArray.Vectorizable.B` was flattened as expected, but `TestArray.Vectorizable.C` gave:
> 
> > Translation Error
> > Internal error Instantiation of TestArray.Vectorizable.C failed with no error message.
> 
> I guess there is still something going wrong with the mergin. Can you please check?

Seems the merging is fine, but `nonfScalarize` has some issues with the resulting model. I'll check it out tomorrow.

**Comment by perost** [comment]
The issue with `TestArray.Vectorizable.C` should be fixed in #8495. It was a general issue with `nonfScalarize` that just happened to show up when introducing arrays in certain contexts.

**Comment by casella** [comment]
I just tried it, looks good 😃.

Only one issue remains: the merged model is equivalent to the original one, except for the variable names, that are losing information because meaningful names are replaced by dummy identifiers build with the class name and an integer. Complex models will become tricky to understand by the user.

It would be nice if we could also get some kind of Rosetta stone, which shows the correspondence between the original instance names and the merged ones. This could be placed in vendor annotation, which could then be analyzed by some post-processing tool (possibly OMEdit itself, one day).

@perost, what do you think?

BTW, I guess this feature could be relevant for the LargeDyn project. I would suggest you to try merging components (no arrayConnect!) in the models of the ScalablePowerGrids library, which are built with large numbers of instances of the same class. The smallest model has about 2000 scalar equations, the largest one has about 3.5 millions. Could you give it a try? I'd like to, but I'm currently swamped.

**Comment by perost** [comment]
> I just tried it, looks good smiley.
> 
> Only one issue remains: the merged model is equivalent to the original one, except for the variable names, that are losing information because meaningful names are replaced by dummy identifiers build with the class name and an integer. Complex models will become tricky to understand by the user.
> 
> It would be nice if we could also get some kind of Rosetta stone, which shows the correspondence between the original instance names and the merged ones. This could be placed in vendor annotation, which could then be analyzed by some post-processing tool (possibly OMEdit itself, one day).
> 
> @perost, what do you think?

I already create a name map in order to replace the names everywhere in the model, so it's just a matter of deciding on what format we want it in.

**Comment by perost** [comment]
@casella: #8505 adds dumping of the name map using the same JSON format as the model obfuscation feature. Though looking at it now I realize it's reversed compared to what you suggested, i.e. `"oneInstance": "$Model[1]"` instead of `"$Model[1]": "oneInstance"`. It wouldn't be hard to add an option to swap the key and value to the dumping function, but maybe it doesn't really matter?

We now also dump the mapping to file when the frontend is run with the merging enabled, which I don't like since we usually don't generate any files when just flattening a model. I think it would be better to only dump the file if we're doing e.g. `simulate` or `buildModel`, but the frontend doesn't really know what the flat model will actually be used for.

**Comment by casella** [comment]
> @casella: #8505 adds dumping of the name map using the same JSON format as the model obfuscation feature.

Sounds good.

> Though looking at it now I realize it's reversed compared to what you suggested, i.e. `"oneInstance": "$Model[1]"` instead of `"$Model[1]": "oneInstance"`. It wouldn't be hard to add an option to swap the key and value to the dumping function, but maybe it doesn't really matter?

It really doesn't. In the other order the file will probably look tidier to a human eye, because all the $Model[xx] will be aligned, but who cares, this is not meant to be read by humans.

> We now also dump the mapping to file when the frontend is run with the merging enabled, which I don't like since we usually don't generate any files when just flattening a model. I think it would be better to only dump the file if we're doing e.g. `simulate` or `buildModel`, but the frontend doesn't really know what the flat model will actually be used for.

Well, this file is integral part of the information about the flat model, so I don't see why it shouldn't be generated all the time a model is flattened. In fact, our LLVM compiler uses the output of instantiateModel() as a starting point, if you don't dump that file, that information would be missing. Also in the case you want to look at the instantiated model in OMEdit, you will probably want to know which variable is which. The only case when that file is not really useful is when doing checkModel, but I don't see that as a problem.

**Comment by perost** [comment]
Some more benchmarking using `ScalableTestGrids`:
| Model                    |       Variables | Without Merging (s) | With Merging (s) | 
| :----------------------- | -----------------:| --------------: | -----------: |
| Type1_N_1_M_1   |             1920 |          0.31  |         0.24 |
| Type1_N_1_M_2   |             2196 |          0.29  |         0.17 |
| Type1_N_2_M_2   |             9966 |          1.80  |         0.78 |
| Type1_N_2_M_4   |           12174 |          1.87  |         0.77 |
| Type1_N_3_M_4   |           28284 |          4.66  |         2.15 |
| Type1_N_4_M_4   |           51078 |          8.64  |         3.88 |
| Type1_N_6_M_4   |         116718 |          20.5  |         10.6 |
| Type1_N_8_M_4   |         209094 |          36.0  |         22.5 |
| Type1_N_11_M_4 |         397788 |          70.7  |         56.5 |
| Type1_N_16_M_4 |         845958 |        162.4  |       190.0 |
| Type1_N_23_M_4 |       1754124 |        496.6  |       735.9 |
| Type1_N_32_M_4 |          ?           | :skull_and_crossbones: | 2644 |

The last one got through the frontend when using merging but got killed during checkModel, hence why it has a time even though the number of variables is unknown. Without merging it used too much memory to even pass the frontend. 

Something weird is going on for the largest models where merging start to be slower than without (though using much less memory). The issue seems to be that the `NFEvalConstants` phase starts to use a lot of time, so it might have some issues with very large arrays.

**Comment by casella** [comment]
@perost I guess you are reporting the frontend time only, right?

When doing merging, it is very clear there is something in there (probably NFEvalConstants) that scales as O(N^2) and dominates the time spent in the FE when there are significantly more than 200.000 equations. Without that, the N_32_M_4 model would probably take around 300 s. I guess that would be a trifle too much, since the whole cycle time for TSOs is 15 minutes, and we can't spend 5 of them just with flattening. But it would already be much better than 2644 s.

**Comment by perost** [comment]
> @perost I guess you are reporting the frontend time only, right?

Yes

> When doing merging, it is very clear there is something in there (probably NFEvalConstants) that scales as O(N^2) and dominates the time spent in the FE when there are significantly more than 200.000 equations. Without that, the N_32_M_4 model would probably take around 300 s. I guess that would be a trifle too much, since the whole cycle time for TSOs is 15 minutes, and we can't spend 5 of them just with flattening. But it would already be much better than 2644 s.

1705s is spent on NFEvalConstants, 797s on NFFlatten. So there's certainly a lot of room for improvements there.

**Comment by casella** [comment]
@perost, is there any debugging technique out there to identify double nested loops (hence scaling as O(N^2)) at runtime?

**Comment by perost** [comment]
> @perost, is there any debugging technique out there to identify double nested loops (hence scaling as O(N^2)) at runtime?

Not that I'm aware of. Finding nested for-loops would be quite easy, but performance issues are almost never that obvious in my experience but rather linear functions calling other linear functions (often indirectly).

But anyway, I've already fixed the worst performance issues for this case. The N_11_M_4 model now takes 27s while the N_23_M_4 model takes 123s, so much better scaling. But I'm still working on implementing the fix properly, MetaModelica is being difficult as usual.

Also, I found out that the numbers are not accurate the way I did it, because I just put all the `checkModel` commands in the same script. This causes the GC to have memory preallocated when flattening the next model in the list, which heavily influences the performance. So for example the N_11_M_4 model that took 70.7s (without merging) actually takes around 112s when flattening it "cold".

**Comment by casella** [comment]
> Not that I'm aware of. Finding nested for-loops would be quite easy, but performance issues are almost never that obvious in my experience but rather linear functions calling other linear functions (often indirectly).

Sure, that's what I meant for runtime double-nested loops :)

Maybe looking at the stack trace after a few hundred seconds of running and trying to find patterns among the function calls could help?

> But anyway, I've already fixed the worst performance issues for this case. The N_11_M_4 model now takes 27s while the N_23_M_4 model takes 123s, so much better scaling.

Yes, that looks pretty linear, which is already much better than O(N^2). However, I wonder, are you doing the constant evaluation and the flattening before or after merging? In the former case, shouldn't this task have O(N) performance, at least some parts of it?

Anyway, we don't have a strong need to make it faster than that right now, maybe later on as we manage to actually run some simulations using array-preserving code generation, if that turns out to be a major bottleneck.

> But I'm still working on implementing the fix properly, MetaModelica is being difficult as usual.

Good luck with that :)

> Also, I found out that the numbers are not accurate the way I did it, because I just put all the `checkModel` commands in the same script. This causes the GC to have memory preallocated when flattening the next model in the list, which heavily influences the performance. So for example the N_11_M_4 model that took 70.7s (without merging) actually takes around 112s when flattening it "cold".

Yeah, memory-management-related performance issues are tricky.

**Comment by perost** [comment]
> Yes, that looks pretty linear, which is already much better than O(N^2). However, I wonder, are you doing the constant evaluation and the flattening before or after merging? In the former case, shouldn't this task have O(N) performance, at least some parts of it?

This is with scalarization, turning off scalarization doesn't work for these models for some reason. I will see if I can fix that too.

**Comment by casella** [comment]
Aha, of course we should have turned that off :)

Good luck with that also. I really expect those times to go down substantially once you have -nonfScalarize working.

**Comment by casella** [comment]
@perost, as @kabdelhak is getting closer to be able to handle Complex operator records in the new backend, we need to make sure that the new frontend can deliver on large-scale power grids models. 

I made some further tests with this simple test script on my i9 12900KF

```
loadModel(ScalableTestGrids); getErrorString();
setCommandLineOptions("-d=mergeComponents,evaluateAllParameters,vectorizeBindings --newBackend  ");getErrorString();
instantiateModel(ScalableTestGrids.Models.Type1.Type1_N_8_M_4);getErrorString();
```
which is supposed to merge scalar components into large arrays and flatten then avoiding scalarization (thanks to `--newBackend`). Notice that there will be a large numbers of individual (scalar) connect statements that cannot be merged into array-equations.

I obtained the following results:

| Model                    |       Variables | flatten (s) |evaluate (s) | simplify (s) | total NF time (s) | 
| :--------------------| --------------:| -----------: | -----------: | -----------: |-----------: |
| Type1_N_4_M_4   |           51078 |         0.8 |      0.3 |       0.2 |     1.9 |
| Type1_N_6_M_4   |         116718 |         1.8 |      1.8 |       0.8 |     4.8 |
| Type1_N_8_M_4   |         209094 |         4.8 |      5.1 |       2.5 |   13.4 |
| Type1_N_11_M_4   |       397788 |       14.3 |     17.2|       8.7 |   42.0 |
| Type1_N_16_M_4   |        845958|       62.4 |    80.1 |     42.8 | 180.7 |
| Type1_N_23_M_4   |      1754124|     251.4 |  323.6 |   172.2 | 750.0 |

There is clearly something not scaling up properly. The results you reported back in March seemed to scale in a slightly superlinear fashion, but now I see figures blowing up quadratically (or more). Has something changed since then? Maybe you could try with older versions and do some bisection to figure out.

This is really a strategic issue. Our goal is to be able to build and simulate power grids models with several millions equations in less than 15 minutes. We can't allocate more than 100-200 seconds to the frontend. I was optimistic that this could be achieved, since after merging, all operations on models should be carried out only once for the entire array, so I was expecting flattening times of a few tens of seconds at most.

I was worried about the connect statements, which are not collected into arrays, but NFFlatten.resolveConnections takes less than 1 s even in the larger models, so it looks like that is not an issue. As to NFInstUtil.mergeScalars, it takes less than 0.1 s, so that's no problem either.

**Comment by casella** [comment]
In the flattened models, I see a lot of parameter array declarations with very large literal arrays of equal values, I guess because those values are not changed, so the default binding is kept. Could it be that this is what causes the time to blow up? In that case, could you use fill(dim1, dim2, ..., default_value) for the RHS of the binding, instead of generating those huge dumb arrays filled with the same value over and over?

**Comment by casella** [comment]
It looks like the story is more complicated.

I checked the source code of the model, and indeed modifiers are explicitly given for most public parameters when instantiating individual components - they just happen to be the same for all of them. Ditto for the many final parameters with binding equations depending on them.

Although this may be the case for some parameters in real applications, chances are that in real life they won't - most of them will probably have a value chosen in a small set, but it's unlikely that they are all equal. For example, the nominal voltages of lines may have two values, corresponding to HV ad EHV lines. So, figuring out a posteriori that they are all the same and using compact notation such as `fill()` could be done, but won't probably be very useful in real applications.

I think what is really necessary is to figure out what is the root cause of `O(N^a)` scaling with `a >= 2` and remove it. I can't see any reason why something like this shouldn't scale as `O(N)`, or maybe as `O(Nlog(N))` in the worst case.

**Comment by perost** [comment]
#10808 more than halves the time for simplification by not trying to simplify arrays that only contain literal values, which isn't even low-hanging fruit but fruit laying on the ground. I will continue to investigate this though, the fact that the evaluation phase is so slow indicates that there might be large arrays of literals which are not marked as such for some reason.

**Comment by casella** [comment]
> #10808 more than halves the time for simplification by not trying to simplify arrays that only contain literal values, which isn't even low-hanging fruit but fruit laying on the ground.

Good.

> I will continue to investigate this though, the fact that the evaluation phase is so slow indicates that there might be large arrays of literals which are not marked as such for some reason.

I guess so. There's stuff scaling badly, so I guess some simple profiling on the largest example should point to the culprit(s).

**Comment by perost** [comment]
I made two more optimizations in #10810 and #10813. The first is to avoid mapping over arrays that only contain literal values, because we do a lot of mapping but never to manipulate literal values. 

And the second is that we do the simplification `{1, 1, 1, 1, 1, 1}[i] => 1`, but this requires us to check that all the elements in the array are equal. These models are kind of a worst case for this, because they contain huge arrays that require us to go through a large part of the elements to find out that they're not actually all equal. But in this case the iterators are actually simplified to literals too, in which case there's no point in checking if the elements are equal or not.

Together this leads to some decent improvements, here are the results on my 3900X (which is a fair bit slower than your 12900KF):
| Model                    |       Variables | flatten (s) |evaluate (s) | simplify (s) | total NF time (s) | 
| :--------------------| --------------:| -----------: | -----------: | -----------: |-----------: |
| Type1_N_11_M_4   |       397788 |       1.0 |     1.4|       0.7 |   4.2 |
| Type1_N_16_M_4   |        845958|       2.4 |    3.3 |    1.6 | 9.1 |
| Type1_N_23_M_4   |      1754124|     4.9 |  6.9 |   1.9 | 17.7 |
| Type1_N_32_M_4   |      3403014|     8.4 |  13.5 |   4.4 | 35.3 |

Note that I added the N_32_M_4 model last since it's now fast enough to test. I haven't investigated further yet, but my intuition says that the evaluate step seems a bit slower than I would expect. The simplify step could also be improved, but flatten might be hard to do much about.

**Comment by casella** [comment]
@perost tomorrow I'll check this with the nightly build on my server. This looks pretty good already: 35 s fo 3.4 M equations is acceptable, and most importantly I see linear scaling, so if we go to larger models it won't take forever.

I can't wait when we'll be able to run Type1_N_32_M_4 through the new backend. This model is already representative of the size of the French grid, and the two larger ones of pan-european ENTSO-E system models. 

BTW, what is really good now is that the Object-Oriented structure is no longer a major overhead: alias variable and similar can be elminated wholesale on arrays, so stripping down the 3.4 M equations to the set of non-trivial equations, probably between 500k and 1M will be done very fast.

**Comment by casella** [comment]
Linux build is broken since Jun 10, I'll test it as soon as it is available again.

**Comment by casella** [comment]
@perost, here are the results with the i9 server after your improvements

| Model                    |       Variables | flatten (s) |evaluate (s) | simplify (s) | total NF time (s) | 
| :--------------------| --------------:| -----------: | -----------: | -----------: |-----------: |
| Type1_N_4_M_4   |           51078 |         0.1 |      0.7 |       0.0 |     1.1 |
| Type1_N_6_M_4   |         116718 |         0.7 |      0.2 |       0.1 |     1.7 |
| Type1_N_8_M_4   |         209094 |         0.8 |      0.7 |       0.1 |     2.4 |
| Type1_N_11_M_4   |       397788 |         1.0 |      0.9 |       1.5 |     4.3 |
| Type1_N_16_M_4   |        845958|         2.0 |      2.5 |       1.1 |     7.8 |
| Type1_N_23_M_4   |      1754124|         3.3 |      5.2 |       1.7 |   16.2 |
| Type1_N_32_M_4   |      3403014|         6.0 |      9.5 |       3.1 |   30.6 |

Comparable to what you got, only slightly faster, and quite good. Thanks!

**Comment by phannebohm** [comment]
Just read a few of the older comments. For the record, here is an updated link to Dirk Zimmer's article on [Module-Preserving Compilation of Modelica Models](https://ep.liu.se/en/conference-article.aspx?series=ecp&issue=43&Article_No=104).

**Comment by casella** [comment]
Dirk is coming up with a lot of interesting ideas to sidestep issues that have plagued our community for two decades. And, he does it on the modelling side, which is really cool. 

He's got two papers and a plenary at the next Conference, I guess it will be the most interesting stuff of this edition 😄

**Comment by casella** [comment]
I got back here while reviewing old tickets. I think the results posted [here](https://github.com/OpenModelica/OpenModelica/issues/3678#issuecomment-1590199955) are quite good already and that the frontend will probably not be the bottleneck for 3.4M equation models for quite some time, so I'm closing this ticket.

@phannebohm, at some point we may also think again about what @dzimmer wrote in 2009. Food for thought.
